src/main/scala/util/PerformanceTimer.scala

```diff
@@ -2,6 +2,30 @@ package util
 import scala.collection.mutable
 import scala.collection
 
+case class RegionTimer(name: String) {
+  private var total: Long = 0
+  private var entered: Long = 0
+  private var inside: Boolean = false
+
+  def enter() = {
+    require(!inside)
+    inside = true
+    entered = System.currentTimeMillis()
+  }
+
+  def exit() = {
+    require(inside)
+    inside = false
+    total += (System.currentTimeMillis() - entered)
+  }
```

Comment:
you can have a method which takes a `=> T` parameter. this would allow you to write code like
```scala
region.within {
  // code to be timed
}
```

```suggestion
  def within[T](body: => T): T = {
    enter()
    val result = body
    exit()
    result
  }

  def enter() = {
    require(!inside)
    inside = true
    entered = System.currentTimeMillis()
  }

  def exit() = {
    require(inside)
    inside = false
    total += (System.currentTimeMillis() - entered)
  }
```